src/main/java/uk/gov/cslearning/record/service/CourseRefreshService.java

```diff
@@ -1,92 +1,150 @@
+            log.error("Splitting user {}", userId);
+            if (userSplit.containsKey(userId)) {
+                userSplit.get(userId).add(statement);
+            } else {
+                ArrayList<Statement> newUserStatements = new ArrayList();
+                newUserStatements.add(statement);
+                userSplit.put(userId, new ArrayList<>(newUserStatements));
             }
+        }
 
-            Collection<CourseRecord> updatedCourseRecords = new ArrayList();
+        Collection<CourseRecord> updatedCourseRecords = new ArrayList();
+
+        for (String userId : userSplit.keySet()) {
+            Collection<CourseRecord> existingCourseRecords = transactionTemplate.execute(status -> courseRecordRepository.findByUserId(userId));
+            Collection<CourseRecord> userRecords = stream.replay(userSplit.get(userId), statement -> ((Activity) statement.getObject()).getId(), existingCourseRecords);
+            updatedCourseRecords.addAll(userRecords);
+        }
 
-            for (String userId : userSplit.keySet()) {
-                Collection<CourseRecord> existingCourseRecords = transactionTemplate.execute(status -> courseRecordRepository.findByUserId(userId));
-                Collection<CourseRecord> userRecords = stream.replay(userSplit.get(userId), statement -> ((Activity) statement.getObject()).getId(), existingCourseRecords);
-                updatedCourseRecords.addAll(userRecords);
+        transactionTemplate.execute(new TransactionCallbackWithoutResult() {
+            @Override
+            protected void doInTransactionWithoutResult(TransactionStatus status) {
+                courseRecordRepository.saveAll(updatedCourseRecords);
             }
+        });
+        log.info("Statements saved to DB");
+        return updatedCourseRecords.size();
+    }
+
+    private List<Statement> fetchStatementsBySinceDate(LocalDateTime since) {
+        MongoDatabase mongoDatabase = mongoClient.getDatabase(mongoProperties.getDatabase());
+        MongoCollection<Document> documents = mongoDatabase
+            .getCollection(MONGO_DOCUMENT_NAME);
+        int lastPageNumber = calculateLastPage(documents);
+        List<Statement> statements = new ArrayList<>();
 
-            transactionTemplate.execute(new TransactionCallbackWithoutResult() {
-                @Override
-                protected void doInTransactionWithoutResult(TransactionStatus status) {
-                    courseRecordRepository.saveAll(updatedCourseRecords);
-                }
-            });
-            log.info("Statements saved to DB");
-            return updatedCourseRecords.size();
-        } catch (IOException e) {
-            throw new RuntimeException("Exception retrieving xAPI statements.", e);
+        for (int pageNum = 0; pageNum < lastPageNumber; pageNum++) {
+            statements.addAll(documents.find(gte(TIMESTAMP_FIELD, Date.from(since.atZone(ZoneId.systemDefault()).toInstant())))
```

Comment:
Can we check if there's an index worth adding on the statements collection to help this? Seeing as we're potentially adding a set of indexes for the 3.6 upgrade anyhow.